CLDC-1945 numeric questions decimal letter entry

arfacamble opened the pull request:
https://digital.dclg.gov.uk/jira/browse/CLDC-1945

This ticket is to deal with the situation currently where questions relating to integer fields in the db would accept decimals and round/truncate them silently.
The user will now see a validation indicating the degree of accuracy required for the question.

Relatedly users are able to enter the letter 'e' in number input fields, since HTML allows representing numbers in exponent form (eg 1e2 = 1 * 10^2) - users who enter an e in an invalid exponent form (eg 'e' or 'e1') we will allow HTML to treat that as invalid, empty the field and the standard question is required validation will show. users who enter a number in valid exponent form (eg '1e2') will see a validation asking them to enter the number in normal number digits form

|<img width="823" alt="image" src="https://user-images.githubusercontent.com/51094020/226317690-a529d0d4-0913-4fc3-92fc-7725701e9d39.png">
|-

---

N.B.
the values for @step for each question have been checked and cleared with Jemma. Worth noting that several fields representing currency are decimal with scale 2 in the db but we are now only allowing integer. There may be "invalid" values in the db and we may want to do tech debt to change the column type but I imagine both of these things are really low priority

arfacamble committed:
rename variables in test file

arfacamble committed:
write tests for new validations

arfacamble committed:
validate that answers to numeric questions are given to the correct level of accuracy and in normal format with only digits

arfacamble committed:
add error messages for new validations

arfacamble committed:
add or alter step on numeric questions in sales, amend one test given step changes

arfacamble committed:
copy change

arfacamble commented on spec/models/validations/shared_validations_spec.rb:
apologies there is a lot of renaming diff here.
I will comment in line where actual changes are in this file and please ignore everything else

arfacamble commented on spec/models/validations/shared_validations_spec.rb:
```diff
@@ -15,119 +15,177 @@
       end
     end
   end
 
```
HERE ARE THE CHANGES!!!!!

kosiakkatrina commented on app/models/validations/shared_validations.rb:
```diff
@@ -34,6 +34,35 @@ def validate_numeric_min_max(record)
+  def validate_numeric_step(record)
+    record.form.numeric_questions.each do |question|
+      next unless question.step
+      next unless record[question.id] && question.page.routed_to?(record, nil)
```
Wouldn't the value get cleared if the page is not routed to?

arfacamble replied:
hmm. I might try and have a little experiment with this. What you say makes sense, I wonder if there's a chance because we're in a before-save lifecycle call back if there might be funny edge cases.
Really here, I copied the format of the above method for validate_numeric_min_max so I could get started and didn't come back round to question this.
Now I look in git blame it looks like you last touched the corresponding line on l21 above, maybe you can enlighten me as to why I need to do this! I will see what happens if I get rid of it

arfacamble replied:
hmmm. I am happy to remove this if you think it's fine, I wonder if there's an edge case where it makes sense though.
A page can be not routed to if the subsection it is in is not enabled, and perhaps that answer could be derived? I should hope that we don't derive invalid values, but maybe it's not crazy to have an extra safeguard that a validation error gets added to a question that the user can't even get to? although in that case we will be calling a value from the db that has already passed this validation, so I really can't imagine a case where this is necessary.

OK I see that you did add it in the other method and discussed it with sam in review
https://github.com/communitiesuk/submit-social-housing-lettings-and-sales-data/pull/1249/files#diff-86ebd101437a30c8df76e2e300c9c61a9cdecefd09e7594fe02cad46e7234aa2
maybe when you're free you can call/message if you remember that thing and whether it applies here (I don't tihnk I have the context for that conversation)

kosiakkatrina commented on app/models/validations/shared_validations.rb:
```diff
@@ -34,6 +34,35 @@ def validate_numeric_min_max(record)
+        answer = Float(input)
+      rescue ArgumentError
+        add_step_error(record, question)
+      end
```
is there a test for this?

jamesrose commented on app/models/validations/shared_validations.rb:
```diff
@@ -34,6 +34,35 @@ def validate_numeric_min_max(record)
+        record.errors.add question.id.to_sym, I18n.t("validations.numeric.normal_format")
+      end
+    end
+  end
```
I don't think we should do this. Let's discuss!

jamesrose commented on config/locales/en.yml:
```diff
@@ -139,6 +139,11 @@ en:
       above_min: "%{field} must be at least %{min}"
+      whole_number: "%{field} must be a whole number"
+      nearest_ten: "%{field} must be given to the nearest ten"
+      nearest_penny: "%{field} must be given to the nearest penny"
```
Are you sure this is always a monetary amount?

arfacamble replied:
I did check that currently all questions with step 0.01 are representing currency, but I agree this made me pause

jamesrose replied:
I think we need to keep in mind that numeric fields aren't limited to currency, and might not be in the future. Suggest changing `nearest_penny` to something that works for all numeric types.

arfacamble committed:
remove validation that is specific to number format and fold that functionality into accuracy validation, rejig tests accordingly

arfacamble committed:
fix various tests

arfacamble committed:
remove reference to currency in validation message that concerns step only and doesn't check whether firled is currency related

arfacamble committed:
alter copy to avoid redundant repetition in validaiton message

arfacamble committed:
save changes made in rebase

arfacamble committed:
changes after PO review
two step values had been missed in previous work
in various places there were custom validations applied that duplicated the functionality of the generic min max validation that is applied to all numeric questions, in these situations sometimes the min and max on the question class was inconsistent with the limit that triggered validations
these have been corrected and made consistent
various tests were affected by this and have been amended accordingly

arfacamble committed:
remove tests introduced by mistake in rebase
